### PR TITLE
[Feature][Executor] Enable batch timeout for python flow batch run

### PR DESCRIPTION
**Pull request by PeiwenGaoMS**
# Description

Enable batch timeout for python flow batch run.

This pull request primarily introduces a batch timeout feature to the `promptflow` package, along with some code refactoring for improved readability and performance. The batch timeout feature allows setting a maximum time limit for batch execution, which can help prevent long-running processes from consuming resources indefinitely. The refactoring includes changes to the handling of exceptions and the use of timeouts.

Here are the most important changes:

Batch Timeout Feature:
* [`src/promptflow/promptflow/batch/_batch_engine.py`](diffhunk://#diff-ecf0905f2116abe08b5cf2931b856bf39aa8b38a30fadd9042538d076dbfde80R83): Added a new `batch_timeout_sec` parameter to the `BatchEngine` class, which sets a maximum time limit for batch execution. This timeout value can either be provided directly or fetched from an environment variable. [[1]](diffhunk://#diff-ecf0905f2116abe08b5cf2931b856bf39aa8b38a30fadd9042538d076dbfde80R83) [[2]](diffhunk://#diff-ecf0905f2116abe08b5cf2931b856bf39aa8b38a30fadd9042538d076dbfde80L112-R126)
* [`src/promptflow/promptflow/batch/_python_executor_proxy.py`](diffhunk://#diff-8c42522c7186705de9a6651343af6006be02b7342f72467c2785f23f93ea669eR55-R84): Modified the `_exec_batch` method to accept `batch_timeout_sec` and `line_timeout_sec` parameters. These parameters are then used to set timeouts for the batch and individual lines, respectively.
* [`src/promptflow/promptflow/executor/_errors.py`](diffhunk://#diff-1f02407850413baccbb33e0078d9772e15ae1c46c020dc7457529c6f7a054315L169-R185): Introduced a new `BatchExecutionTimeoutError` exception that gets raised when the batch timeout is exceeded.
* [`src/promptflow/promptflow/executor/_line_execution_process_pool.py`](diffhunk://#diff-c14422c33faa9e44145738152120929455da4e4f05d4fc7b59158423fa609bc1L263-R271): Added checks for batch and line timeouts in the `_monitor_workers_and_process_tasks_in_thread` method. If a timeout is detected, the corresponding exception is raised. [[1]](diffhunk://#diff-c14422c33faa9e44145738152120929455da4e4f05d4fc7b59158423fa609bc1L263-R271) [[2]](diffhunk://#diff-c14422c33faa9e44145738152120929455da4e4f05d4fc7b59158423fa609bc1L285-L287) [[3]](diffhunk://#diff-c14422c33faa9e44145738152120929455da4e4f05d4fc7b59158423fa609bc1L300-R305)

Code Refactoring:
* [`src/promptflow/promptflow/batch/_batch_engine.py`](diffhunk://#diff-ecf0905f2116abe08b5cf2931b856bf39aa8b38a30fadd9042538d076dbfde80L179-R188): Simplified several lines of code for improved readability. For example, the handling of inputs in the `run` method and the execution of lines in the `_exec` method were simplified. [[1]](diffhunk://#diff-ecf0905f2116abe08b5cf2931b856bf39aa8b38a30fadd9042538d076dbfde80L179-R188) [[2]](diffhunk://#diff-ecf0905f2116abe08b5cf2931b856bf39aa8b38a30fadd9042538d076dbfde80L258-R274)
* [`src/promptflow/promptflow/executor/_flow_nodes_scheduler.py`](diffhunk://#diff-33de607e09569e7e33422ca25579a82d0418874a4e917cd711b48dc8c6dd585cL70-R70): Simplified the handling of completed futures in the `execute` method.
* [`src/promptflow/promptflow/executor/_line_execution_process_pool.py`](diffhunk://#diff-c14422c33faa9e44145738152120929455da4e4f05d4fc7b59158423fa609bc1R98-R101): Removed redundant code related to line timeout and restructured the `_monitor_workers_and_process_tasks_in_thread` method for improved readability. [[1]](diffhunk://#diff-c14422c33faa9e44145738152120929455da4e4f05d4fc7b59158423fa609bc1R98-R101) [[2]](diffhunk://#diff-c14422c33faa9e44145738152120929455da4e4f05d4fc7b59158423fa609bc1L120-R126) [[3]](diffhunk://#diff-c14422c33faa9e44145738152120929455da4e4f05d4fc7b59158423fa609bc1L236-R245)

# All Promptflow Contribution checklist:
- [x] **The pull request does not introduce [breaking changes].**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](../CONTRIBUTING.md).**
- [ ] **Create an issue and link to the pull request to get dedicated review from promptflow team. Learn more: [suggested workflow](../CONTRIBUTING.md#suggested-workflow).**

## General Guidelines and Best Practices
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### Testing Guidelines
- [x] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: src/promptflow/promptflow/batch/_batch_engine.py [hunk 1]
```diff
@@ -99,7 +102,15 @@ def __init__(
 
         self._connections = connections
         self._storage = storage
+
+        # process timeout parameters and pass them to create function of executor proxy
+        self._batch_timeout_sec = (
+            batch_timeout_sec if batch_timeout_sec else get_int_env_var("PF_BATCH_TIMEOUT_SEC", None)
+        )
+        self._line_timeout_sec = get_int_env_var("PF_LINE_TIMEOUT_SEC", LINE_TIMEOUT_SEC)
+        kwargs.update({"line_timeout_sec": self._line_timeout_sec})
```
Comment (thy09) [inline src/promptflow/promptflow/batch/_batch_engine.py:111]: We may add some comment why we add this to kwargs?

File: src/promptflow/promptflow/batch/_python_executor_proxy.py [hunk 2]
```diff
@@ -28,8 +29,15 @@ async def create(
         storage: Optional[AbstractRunStorage] = None,
         **kwargs,
     ) -> "PythonExecutorProxy":
-        # TODO: Raise error if connections is None
-        flow_executor = FlowExecutor.create(flow_file, connections, working_dir, storage=storage, raise_ex=False)
+        line_timeout_sec = kwargs.get("line_timeout_sec", LINE_TIMEOUT_SEC)
```
Comment (thy09) [inline src/promptflow/promptflow/batch/_python_executor_proxy.py:32]: Should we add the parameter here?

File: src/promptflow/promptflow/batch/_python_executor_proxy.py [hunk 3]
```diff
@@ -46,12 +54,13 @@ def _exec_batch(
         batch_inputs: List[Mapping[str, Any]],
         output_dir: Path,
         run_id: Optional[str] = None,
+        batch_timeout_sec: Optional[int] = None,
     ) -> List[LineResult]:
         self._flow_executor._node_concurrency = DEFAULT_CONCURRENCY_BULK
         with self._flow_executor._run_tracker.node_log_manager:
             OperationContext.get_instance().run_mode = RunMode.Batch.name
             line_results = self._flow_executor._exec_batch_with_process_pool(
```
Comment (thy09) [inline src/promptflow/promptflow/batch/_python_executor_proxy.py:62]: Can we decouple exec_batch_with_process_pool and flow_executor

File: src/promptflow/promptflow/executor/_line_execution_process_pool.py [hunk 4]
```diff
@@ -328,6 +331,15 @@ def _monitor_workers_and_process_tasks_in_thread(
 
             self._processing_idx.pop(line_number)
 
+        # End the process when the batch timeout is exceeded.
+        self._processes_manager.end_process(index)
+        self._ensure_process_terminated_within_timeout(process_id)
+
+    def _check_batch_within_timeout(self, start_time: datetime) -> bool:
```
Comment (thy09) [inline src/promptflow/promptflow/executor/_line_execution_process_pool.py:338]: Maybe `_timeout_expired`?

Comment (PeiwenGaoMS) [reply src/promptflow/promptflow/executor/_line_execution_process_pool.py]: Updated.

File: src/promptflow/promptflow/batch/_python_executor_proxy.py (hunk 2 above)
Comment (PeiwenGaoMS) [reply src/promptflow/promptflow/batch/_python_executor_proxy.py]: Updated. Moved to exec_batch.

File: src/promptflow/promptflow/batch/_batch_engine.py (hunk 1 above)
Comment (PeiwenGaoMS) [reply src/promptflow/promptflow/batch/_batch_engine.py]: Removed.

File: src/promptflow/promptflow/executor/_errors.py [hunk 5]
```diff
@@ -166,9 +166,23 @@ class LineExecutionTimeoutError(UserErrorException):
...
+            target=ErrorTarget.EXECUTOR,
+        )
+
+
+class BatchExecutionTimeoutError(UserErrorException):
+    """Exception raised when batch timeout is exceeded"""
+
+    def __init__(self, line_number, timeout):
+        super().__init__(
+            message_format="Processing of line {line_number} exceeded the batch timeout of {timeout} seconds.",
```
Comment (thy09) [inline src/promptflow/promptflow/executor/_errors.py:182]: BatchExecutionTimeout seems not related to line_number?

Comment (PeiwenGaoMS) [reply src/promptflow/promptflow/executor/_errors.py:182]: Yes. Just to show which line is affected by the batch timeout when executed.

File: src/promptflow/promptflow/batch/_python_executor_proxy.py [hunk 6]
```diff
@@ -51,16 +52,36 @@ def _exec_batch(
...
         with run_tracker.node_log_manager:
             OperationContext.get_instance().run_mode = RunMode.Batch.name
-            line_results = self._flow_executor._exec_batch_with_process_pool(batch_inputs, run_id, output_dir)
+            if self._flow_executor._flow_file is None:
+                raise UnexpectedError(
+                    "Unexpected error occurred while init FlowExecutor. Error details: flow file is missing."
+                )
+
+            if batch_timeout_sec:
+                bulk_logger.info(f"The timeout for batch run is {batch_timeout_sec} seconds.")
```
Comment (thy09) [inline src/promptflow/promptflow/batch/_python_executor_proxy.py:72]: `The timeout for the batch run`

File: src/promptflow/promptflow/executor/_errors.py (hunk 5 above)
Comment (thy09) [reply src/promptflow/promptflow/executor/_errors.py:182]: But this sentence might be confusing, looks like one line exceeded 10000 seconds?

File: src/promptflow/promptflow/batch/_python_executor_proxy.py (hunk 6 above)
Comment (PeiwenGaoMS) [reply src/promptflow/promptflow/batch/_python_executor_proxy.py:72]: Updated.

File: src/promptflow/promptflow/executor/_line_execution_process_pool.py [hunk 7]
```diff
@@ -297,9 +294,15 @@ def _monitor_workers_and_process_tasks_in_thread(
...
-                elif timeouted:
-                    bulk_logger.warning(f"Line {line_number} timeout after {timeout_time} seconds.")
-                    ex = LineExecutionTimeoutError(line_number, timeout_time)
+                elif self._line_timeout_expired(start_time):
+                    bulk_logger.warning(f"Line {line_number} timeout after {self._line_timeout_sec} seconds.")
+                    ex = LineExecutionTimeoutError(line_number, self._line_timeout_sec)
+                # Handle batch execution timeout.
+                elif self._batch_timeout_expired(batch_start_time):
+                    bulk_logger.warning(
+                        f"Line {line_number} execution exceeded the batch timeout of {self._batch_timeout_sec} seconds."
```
Comment (thy09) [inline src/promptflow/promptflow/executor/_line_execution_process_pool.py:301]: Similar problem to the error, might be confusing.

File: src/promptflow/promptflow/executor/_line_execution_process_pool.py [hunk 8]
```diff
@@ -94,6 +95,10 @@ def __init__(
         nlines,
         run_id,
         output_dir,
+        variant_id: str = "",
```
[the remark this answers is not in the document]
Comment (PeiwenGaoMS) [reply src/promptflow/promptflow/executor/_line_execution_process_pool.py]: Removed.

File: src/promptflow/promptflow/executor/_line_execution_process_pool.py [hunk 9]
```diff
@@ -330,6 +333,21 @@ def _monitor_workers_and_process_tasks_in_thread(
 
             self._processing_idx.pop(line_number)
 
+        # End the process when the batch timeout is exceeded.
+        self._processes_manager.end_process(index)
```
Comment (thy09) [inline src/promptflow/promptflow/executor/_line_execution_process_pool.py:335]: We have already stopped them in the loop?

File: src/promptflow/tests/executor/e2etests/test_batch_timeout.py [hunk 10]
```diff
@@ -126,20 +118,48 @@ def test_batch_with_one_line_timeout(self, flow_folder, dev_connections):
...
-        storage: Optional[AbstractRunStorage] = None,
-        **kwargs,
-    ) -> "MockPythonExecutorProxy":
-        line_timeout_sec = kwargs.get("line_timeout_sec", LINE_TIMEOUT_SEC)
-        flow_executor = FlowExecutor.create(
-            flow_file, connections, working_dir, storage=storage, raise_ex=False, line_timeout_sec=line_timeout_sec
+        # assert the error summary in batch result
+        assert batch_results.error_summary.failed_user_error_lines == 1
+        assert batch_results.error_summary.failed_system_error_lines == 0
+        assert isinstance(batch_results.error_summary.error_list[0], LineError)
```
Comment (thy09) [inline src/promptflow/tests/executor/e2etests/test_batch_timeout.py:153]: Check all items in the error summary?

File: src/promptflow/tests/executor/unittests/processpool/test_line_execution_process_pool.py [hunk 11]
```diff
@@ -150,11 +147,7 @@ def execute_in_spawn_mode_subprocess(
 
 
 def create_line_execution_process_pool(dev_connections):
-    executor = FlowExecutor.create(
-        get_yaml_file(SAMPLE_FLOW),
-        dev_connections,
-        line_timeout_sec=1,
-    )
+    executor = FlowExecutor.create(get_yaml_file(SAMPLE_FLOW), dev_connections)
```
Comment (thy09) [inline src/promptflow/tests/executor/unittests/processpool/test_line_execution_process_pool.py:150]: We may change to use other flows instead of sample flow, maybe in another PR.

File: src/promptflow/promptflow/executor/_line_execution_process_pool.py (hunk 9 above)
Comment (PeiwenGaoMS) [reply src/promptflow/promptflow/executor/_line_execution_process_pool.py:336]: Updated.

File: src/promptflow/tests/executor/unittests/processpool/test_line_execution_process_pool.py (hunk 11 above)
Comment (PeiwenGaoMS) [reply src/promptflow/tests/executor/unittests/processpool/test_line_execution_process_pool.py:150]: ok

**Review by thy09 — APPROVED** [review]

**Review by Hhhilulu — APPROVED** [review]

File: src/promptflow/promptflow/batch/_batch_engine.py [hunk 12]
```diff
@@ -109,14 +112,18 @@ def __init__(
...
+            raise InvalidFlowFileError(message_format="Unsupported flow file type: {flow_file}.", flow_file=flow_file)
 
         self._connections = connections
         self._entry = entry
         self._storage = storage
         self._kwargs = kwargs
+
+        self._batch_timeout_sec = (
+            batch_timeout_sec if batch_timeout_sec else get_int_env_var("PF_BATCH_TIMEOUT_SEC", None)
+        )
```
[the remark this answers is not in the document]
Comment (PeiwenGaoMS) [reply src/promptflow/promptflow/batch/_batch_engine.py:124]: will update here in the next PR.
